File: src/main/java/club/libridge/libridgebackend/core/OddTricks.java
package club.libridge.libridgebackend.core;

import java.util.HashMap;
import java.util.Map;

public enum OddTricks {

    ONE("One", "1", 1), TWO("Two", "2", 2), THREE("Three", "3", 3), FOUR("Four", "4", 4), FIVE("Five", "5", 5), SIX("Six", "6", 6),
    SEVEN("Seven", "7", 7);

    private final String name;
    private final String symbol;
    private final int level;

    private static Map<Integer, OddTricks> mapFromLevel = new HashMap<Integer, OddTricks>();
    // Static initialization block to avoid doing this calculation every time
    static {
        for (OddTricks oddTrick : OddTricks.values()) {
            mapFromLevel.put(oddTrick.getLevel(), oddTrick);
        }
    }

    OddTricks(String name, String symbol, int level) {
        this.name = name;
        this.symbol = symbol;
        this.level = level;
    }

    public String getName() {
        return this.name;
    }

    public String getSymbol() {
        return this.symbol;
    }

    public int getLevel() {
        return this.level;
    }

    public static OddTricks fromLevel(int level) {
        if (level < 1 || level > 7) {
            throw new IllegalArgumentException("Level should be between 1 and 7 inclusive.");
        }
        return OddTricks.mapFromLevel.get(Integer.valueOf(level));
    }

}

File: src/main/scala/scalabridge/OddTricks.scala
package scalabridge

import scalabridge.exceptions.OddTricksException
import scala.util.Success
import scala.util.Failure
import scala.util.Try

enum OddTricks(name: String, symbol: String, level: Int) extends java.lang.Enum[OddTricks] {
  case ONE extends OddTricks("One", "1", 1)
  case TWO extends OddTricks("Two", "2", 2)
  case THREE extends OddTricks("Three", "3", 3)
  case FOUR extends OddTricks("Four", "4", 4)
  case FIVE extends OddTricks("Five", "5", 5)
  case SIX extends OddTricks("Six", "6", 6)
  case SEVEN extends OddTricks("Seven", "7", 7)
  def getName: String = this.name
  def getSymbol: String = this.symbol
  def getLevel: Int = this.level
}
object OddTricks {
  def getONE() = ONE
  def getTWO() = TWO
  def getTHREE() = THREE
  def getFOUR() = FOUR
  def getFIVE() = FIVE
  def getSIX() = SIX
  def getSEVEN() = SEVEN
  private val mapFromLevel: Map[Int, OddTricks] =
    OddTricks.values.map(oddTrick => oddTrick.getLevel -> oddTrick).toMap

  def fromLevel(level: Int): Try[OddTricks] = {
    OddTricks.mapFromLevel.get(level) match {
      case Some(oddTrick) => Success(oddTrick)
      case None           => Failure(OddTricksException(level.toString))
    }
  }
}
